pulsar-tracing/src/main/java/com/datastax/oss/pulsar/tracing/BrokerTracing.java

```diff
@@ -0,0 +1,645 @@
+      log.debug(
+          "Invalid tracing level: {}. Setting to NONE for producer {}",
+          producer.getMetadata().get("trace"),
+          producer);
+      return TraceLevel.NONE;
+    }
+  }
+
+  //    private boolean needToTraceTopic(Topic topic) {
+  //        if (topic == null) return false;
+  //
+  //        // todo: how to read topic props
+  //        return true;
+  //    }
+
+  /* ***************************
+   **  Administrative events
+   ******************************/
+
+  public void onConnectionCreated(ServerCnx cnx) {
+    if (!getEnabledEvents(cnx).contains(EventReasons.ADMINISTRATIVE)) return;
+
+    TraceLevel level = getTracingLevel(cnx);
+    if (level == TraceLevel.NONE) return;
+
+    Map<String, Object> traceDetails = new TreeMap<>();
+    populateConnectionDetails(level, cnx, traceDetails);
+    log.info("Connection created: {}", traceDetails);
+  }
+
+  public void producerCreated(ServerCnx cnx, Producer producer, Map<String, String> metadata) {
+    if (!getEnabledEvents(cnx).contains(EventReasons.ADMINISTRATIVE)) return;
+
+    TraceLevel level = getTracingLevel(cnx);
+    if (level == TraceLevel.NONE) return;
+
+    Map<String, Object> traceDetails = new TreeMap<>();
+    populateConnectionDetails(level, cnx, traceDetails);
+    populateProducerDetails(level, producer, traceDetails);
+
+    traceDetails.put("metadata", metadata);
+
+    log.info("Producer created: {}", traceDetails);
+  }
+
+  public void producerClosed(ServerCnx cnx, Producer producer, Map<String, String> metadata) {
+    if (!getEnabledEvents(cnx).contains(EventReasons.ADMINISTRATIVE)) return;
+
+    TraceLevel level = getTracingLevel(cnx);
+    if (level == TraceLevel.NONE) return;
+
+    Map<String, Object> traceDetails = new TreeMap<>();
+    populateConnectionDetails(level, cnx, traceDetails);
+    populateProducerDetails(level, producer, traceDetails);
+
+    traceDetails.put("metadata", metadata);
+
+    log.info("Producer closed: {}", traceDetails);
+  }
+
+  public void consumerCreated(ServerCnx cnx, Consumer consumer, Map<String, String> metadata) {
+    if (!getEnabledEvents(cnx).contains(EventReasons.ADMINISTRATIVE)) return;
+
+    TraceLevel level = getTracingLevel(cnx);
+    if (level == TraceLevel.NONE) return;
+
+    Map<String, Object> traceDetails = new TreeMap<>();
+    populateConnectionDetails(level, cnx, traceDetails);
+    populateConsumerDetails(level, consumer, traceDetails);
+    populateSubscriptionDetails(level, consumer.getSubscription(), traceDetails);
+
+    traceDetails.put("metadata", metadata);
+
+    log.info("Consumer created: {}", traceDetails);
+  }
+
+  public void consumerClosed(ServerCnx cnx, Consumer consumer, Map<String, String> metadata) {
+    if (!getEnabledEvents(cnx).contains(EventReasons.ADMINISTRATIVE)) return;
+
+    TraceLevel level = getTracingLevel(cnx);
+    if (level == TraceLevel.NONE) return;
+
+    Map<String, Object> traceDetails = new TreeMap<>();
```

Comment:
good idea to use a TreeMap, this way the ordering of the key is consistent and they can grep the logs